arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,269 @@
+          TRI_ASSERT(it->size() >= shaIndex + 64);
+          std::string hash = it->substr(shaIndex + /*.sha.*/ 5, 64);
+          it = nextIt;
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          _fileNamesToHashes.try_emplace(sstFileName, std::move(hash));
+        } else {
+          std::string tempPath =
+              basics::FileUtils::buildFilename(_rootPath, *it);
+          LOG_TOPIC("4eac9", DEBUG, arangodb::Logger::ENGINES)
+              << "checkMissingShaFiles:"
+                 " Deleting file "
+              << tempPath;
+          TRI_UnlinkFile(tempPath.data());
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          _fileNamesToHashes.erase(sstFileName);
+        }
+      } else if (isFileNameSst(*it)) {
+        std::unordered_map<std::string, std::string>::const_iterator hashIt;
+        {
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          hashIt = _fileNamesToHashes.find(*it);
+        }
+        if (hashIt == _fileNamesToHashes.end()) {
+          std::string tempPath =
+              basics::FileUtils::buildFilename(_rootPath, *it);
+          LOG_TOPIC("d6c86", DEBUG, arangodb::Logger::ENGINES)
+              << "checkMissingShaFiles:"
+                 " Computing cheacksum for "
+              << tempPath;
+          auto checksumCalc = ChecksumCalculator();
+          if (TRI_ProcessFile(tempPath.c_str(),
+                              [&checksumCalc](char const* buffer, size_t n) {
+                                checksumCalc.updateEVPWithContent(buffer, n);
+                                return true;
+                              })) {
+            checksumCalc.computeFinalChecksum();
+            writeShaFile(tempPath, checksumCalc.getChecksum());
+          }
+        }
+      }
+    }
+  }
+}
+
+rocksdb::Status ChecksumEnv::NewWritableFile(
+    const std::string& fileName, std::unique_ptr<rocksdb::WritableFile>* result,
+    const rocksdb::EnvOptions& options) {
+  std::unique_ptr<rocksdb::WritableFile> writableFile;
+  rocksdb::Status s =
+      rocksdb::EnvWrapper::NewWritableFile(fileName, &writableFile, options);
+  if (!s.ok()) {
+    return s;
+  }
+  try {
+    result->reset(
+        new ChecksumWritableFile(writableFile.release(), fileName, _helper));
+  } catch (std::exception& e) {
+    LOG_TOPIC("8b19e", ERR, arangodb::Logger::ENGINES)
+        << "ChecksumWritableFile: exception caught when allocating "
+        << e.what();
+  }
+
+  return s;
+}
+
+rocksdb::Status ChecksumWritableFile::Append(const rocksdb::Slice& data) {
+  if (_helper->isFileNameSst(_fileName)) {
+    _checksumCalc.updateIncrementalChecksum(data.ToString().data());
+  }
+  return rocksdb::WritableFileWrapper::Append(data);
+}
+
+rocksdb::Status ChecksumWritableFile::Close() {
+  if (_helper->isFileNameSst(_fileName)) {
+    TRI_ASSERT(_helper != nullptr);
+    if (!_helper->writeShaFile(_fileName, _checksumCalc.getChecksum())) {
```

Comment:
When this is called, is the checksum already finalized? I don't see any call to computeFinalChecksum() around here.